### ISSUE TITLE
[networkd] description of UseDNS= / UseDomains=

### ISSUE THREAD

**Issue by msoltyspl** [systemd/systemd#9473]
Systemd version: 238.133
Distro: Archlinux

From manual page: `When true (the default), the DNS servers received from the DHCP server will be used and take precedence over any statically configured ones.`

What I observed is that both sections simply get merged, without any particular preferences (formally DNS= listed DNSes are before dynamically obtained ones). Same applies to Domains= and UseDomains= - static and dynamic are simply merged.

_Probably_ the same applies to NTP= / UseNTP= pair - but I haven't checked this yet.

The behavior is as usual - the relevant DNSes are tried in turn (in case some of servers fail, tested with simple iptables rule)

**Comment by mikhailnov** [comment]
I'm having the same issue. How to disable receiving DNS from DHCP??

```
alt-p8-edu network # cat /etc/resolv.conf
# Generated by resolvconf
# Do not edit manually, use
# /etc/net/ifaces/<interface>/resolv.conf instead.
search lan
nameserver 192.168.1.1
nameserver 8.8.8.8
nameserver 77.88.8.8
alt-p8-edu network # nano *
alt-p8-edu network # cat /etc/systemd/network/*.network
[Match]
Name=en*

[Network]
DHCP=yes
NTP=ntp.ubuntu.com
DNS=8.8.8.8
DNS=77.88.8.8

[DHCP]
UseDNS=false
alt-p8-edu network # 
```

**Comment by ott** [comment]
@mikhailnov UseDNS=false in the DHCP section should be correct.

**Comment by tomtastic** [comment]
```
$ systemctl --version
systemd 240
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

```
$ cat /etc/systemd/networkd.conf
#  This file is part of systemd.
[DHCP]
UseDNS=false
```

```
systemd-networkd[19392]: /etc/systemd/networkd.conf:2: Unknown lvalue 'UseDNS' in section 'DHCP'
```

**Comment by teatreeoilchocolate** [comment]
I stumbled across this issue myself.
I was looking into fixing  a VPN-DNSleak of mine and stumbled across these tickets.
https://github.com/systemd/systemd/issues/6076#issuecomment-429327906
https://github.com/systemd/systemd/issues/10830
https://github.com/systemd/systemd/issues/5755

Since I want my configuration to work for wireless networks outside my home, e.g. coffeshops, I need my setup to be able to adhere their configuration without a lot of fuss. So I thought about, writing a configuration explicitly for the few networks I am using most.
I found that my setup is still acquiring the DNS server configuration via DHCP.
To reproduce the Issue:

1. Create a *.link file in /etc/systemd/network with the appropriate [MATCH] and [NETWORK] section
2. set UseDNS=false under [NETWORK]
3. Disconnect your network
4. execute the following commands as root: systemctl daemon-reload && systemctl restart systemd-networkd && systemctl restart systemd-resolved
5. connect to network listed under [MATCH] section of file created under point 1
6. execute networkctl status and you will find that the network is managed, but also that there is a DNS-configuration present

I assume the reason for this not surfacing sooner is the available alternative of making /etc/resolve.conf a file and not a symlink.
However, this approach has the drawback of not being able to use alternative network-setups with ease, which regularly results in configuration-sessions, rather than getting work done, when entering a new shop trying to use a publicly available WiFi Mac and Windows can use without a problem.

Seeing how the other issues devolved, I want to stress that this is not about the way certain shops run their network or how the VPN-DNS-issue might or might not be an other projects/this projects issue, but rather about the `UseDNS=` configuration not performing actions as documented, thereby limiting flexibility and raising privacy-issues for the user-base.

Following are my system-details:

> % uname -r
5.7.9-arch1-1

>% networkctl status
●         State: routable                                                                      
        Address: 192.168.x.z on wlan0                                                       
                 192.168.x.y on wlan0                                                       
                 172.17.0.1 on docker0                                                 
                 xxxx:xxxx:xxxx:xxxx on wlan0                                            
        Gateway: 192.168.x.x  on wlan0
            DNS: 192.168.x.x # same as gateway                                                                 
  Route Domains: .                                                                             
            NTP: 192.168.x.x # same as gateway

>% cat wlan0_home.network 
[Match]
Name=wlan0
SSID="My Hidden SSID With Spaces"

[Network]
DHCP=ipv4
LLMNR=false
DNSOverTLS=no
DNSSEC=no
[DHCPV4]
UseDNS=false
UseDomains=false 
Anonymize=true

>% resolvectl status
Global
       LLMNR setting: no                  
MulticastDNS setting: no                  
  DNSOverTLS setting: no                  
      DNSSEC setting: allow-downgrade     
    DNSSEC supported: no                  
  Current DNS Server: 9.9.9.9             
         DNS Servers: 9.9.9.9             
Fallback DNS Servers: 1.1.1.1             
                      149.112.112.112     
          DNSSEC NTA: 10.in-addr.arpa     
                      16.172.in-addr.arpa 
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa 
                      18.172.in-addr.arpa 
                      19.172.in-addr.arpa 
                      20.172.in-addr.arpa 
                      21.172.in-addr.arpa 
                      22.172.in-addr.arpa 
                      23.172.in-addr.arpa 
                      24.172.in-addr.arpa 
                      25.172.in-addr.arpa 
                      26.172.in-addr.arpa 
                      27.172.in-addr.arpa 
                      28.172.in-addr.arpa 
                      29.172.in-addr.arpa 
                      30.172.in-addr.arpa 
                      31.172.in-addr.arpa 
                      corp                
                      d.f.ip6.arpa        
                      home                
                      internal            
                      intranet            
                      lan                 
                      local               
                      private             
                      test                
Link 5 (docker0)
      Current Scopes: none           
DefaultRoute setting: no             
       LLMNR setting: yes            
MulticastDNS setting: no             
  DNSOverTLS setting: no             
      DNSSEC setting: allow-downgrade
    DNSSEC supported: yes            
Link 4 (wlan0)
      Current Scopes: DNS          
DefaultRoute setting: yes          
       LLMNR setting: no           
MulticastDNS setting: no           
  DNSOverTLS setting: no           
      DNSSEC setting: no           
    DNSSEC supported: no           
  Current DNS Server: 192.168.x.x
         DNS Servers: 192.168.x.x
Link 2 (enp3s0)
      Current Scopes: none           
DefaultRoute setting: no             
       LLMNR setting: yes            
MulticastDNS setting: no             
  DNSOverTLS setting: no             
      DNSSEC setting: allow-downgrade
    DNSSEC supported: yes

**Comment by keszybz** [comment]
@tomtastic 
> ```
> $ cat /etc/systemd/networkd.conf
> #  This file is part of systemd.
> [DHCP]
> UseDNS=false
> ```
> 
> ```
> systemd-networkd[19392]: /etc/systemd/networkd.conf:2: Unknown lvalue 'UseDNS' in section 'DHCP'
> ```

[DHCP] is a section for .network files. Of course it doesn't work in networkd.conf.

--

I don't see why UseDNS=no wouldn't work. @yuwata added a test for this e2d5aab32afa25d1bb54f4022213edd166199c50. @yuwata do you have any idea why this might not work for them as expected? Were there any fixes that could be relevent?

@msoltyspl, @mikhailnov please tests with a recent version.

**Comment by tomtastic** [comment]
thanks @keszybz, but 2019 was long ago and this is no longer an issue for me :)

**Comment by yuwata** [comment]
> I don't see why UseDNS=no wouldn't work. @yuwata added a test for this [e2d5aab](https://github.com/systemd/systemd/commit/e2d5aab32afa25d1bb54f4022213edd166199c50). @yuwata do you have any idea why this might not work for them as expected? Were there any fixes that could be relevent?

Hmm, I cannot find anything relevant to this issue.
@mikhailnov and @teatreeoilchocolate Could you show the result of `networkctl status INTERFACE_NAME`, and check that .network file name shown in the result is matches the file you set `UseDNS=no`?

**Comment by teatreeoilchocolate** [comment]
Hello!
Sorry for the delay.


This is the output of networkctl status tun0 with UseDNS=no. It is no different from UseDNS=false , according to man-pages it is a boolean anyways. 
```
# networkctl status tun0
● 14: tun0                                                                 
                     Link File: /usr/lib/systemd/network/99-default.link   
                  Network File: /etc/systemd/network/tun0.network          
                          Type: none                                       
                         State: routable (failed)
                        Driver: tun                                        
                           MTU: 1500 (min: 68, max: 65535)                 
                         QDisc: fq_codel                                   
  IPv6 Address Generation Mode: random                                     
          Queue Length (Tx/Rx): 1/1                                        
              Auto negotiation: no                                         
                         Speed: 10Mbps                                     
                        Duplex: full                                       
                          Port: tp                                         
                       Address: xxxxxxx                                 
                                xxxxxxxx                  
                           DNS: 9.9.9.9                                    
                 Route Domains: .                                          

Mar 12 22:37:47 gg systemd-networkd[841]: tun0: Gained IPv6LL
Mar 12 22:37:47 gg systemd-networkd[841]: tun0: DHCP4 CLIENT: Failed to set MAC address: Invalid argument
Mar 12 22:37:47 gg systemd-networkd[841]: tun0: Failed
Mar 12 22:39:34 g systemd-networkd[841]: tun0: Link DOWN
Mar 12 22:39:34 gg systemd-networkd[841]: tun0: Lost carrier
Mar 12 22:39:41 gg systemd-networkd[841]: tun0: Link UP
Mar 12 22:39:41 gg systemd-networkd[841]: tun0: Gained carrier
Mar 12 22:39:41 gg systemd-networkd[841]: tun0: Gained IPv6LL
Mar 12 22:39:41 gg systemd-networkd[841]: tun0: DHCP4 CLIENT: Failed to set MAC address: Invalid argument
Mar 12 22:39:41 gg systemd-networkd[841]: tun0: Failed
```


The DHCP-Problems are peculiar, but do not lead to a defunct device and I doubt they are related to UseDNS.
I also have a tun0.link(under /etc/systemd/network/ which is not used here and ip link lists tun0 as 
` tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN mode DEFAULT group default qlen 500
    link/none`

My configuration-files again because it has been some time.

```
# cat tun0.network 
[Match]
Name=tun0

[Network]
DNS=9.9.9.9
KeepConfiguration=true
Domains=.
DNSSEC=false
DHCP=ipv4
LLMNR=false
[DHCPv4]
UseDNS=false
SendHostname=false
UseDomains=false
Anonymize=true

# cat tun0.link
[Match]
OriginalName=tun0

[Link]
MACAddressPolicy=random

# cat ../networkd.conf
[DHCP]
DUIDType=vendor
DUIDRawData=................
```

**Comment by yuwata** [comment]
@teatreeoilchocolate 
> I also have a tun0.link(under /etc/systemd/network/) which is not used here

You need to prefix a digit to the .link file e.g. 50-tun0.link. We read .network or .link file in alphanumerically, so the file must be listed earlier than 99-default.link.

At least `UseDNS=no/false` issue is not reproducible anymore. Thank you for the confirmation.

> However looking at resolvectl, I can see the search-domain and the DNS-Server for wlan0 being set, BUT the DNS-Server provided via DHCP is the last of the alternative DNS-Servers

I guess you have something like the following config for wlan0
```
[Match]
Name=wlan0

[Network]
DNS=1.1.1.1 8.8.8.8
DHCP=ipv4

[DHCPv4]
UseDNS=yes
```
Hmm, indeed, the man page said
```
UseDNS=
    When true (the default), the DNS servers received from the DHCP server will be used and take precedence over any statically configured ones.
```
If I remember correctly, the behavior was not changed for so long time, so I think we need to adjust the man page.

**Comment by teatreeoilchocolate** [comment]
>You need to prefix a digit to the .link file e.g. 50-tun0.link. We read .network or .link file in alphanumerically, so the file must be listed earlier than 99-default.link.

Thanks. The devil is always in the details.

The wlan0.network file looks indeed similar:
```
[Match]
Name=wlan0
SSID=!"specificSSID"

[Network]
DHCP=ipv4
LLMNR=false
DNSOverTLS=no
DNSSEC=no
[DHCPv4]
UseDNS=true
UseDomains=true
Anonymize=true
```
I take the global configuration for the DNS-servers.
The verbosity stems from another configuration I have for when I do not need  to log into a wireless lan via sign-up-page, and I tried to mitigate the use of the signup-button presented by Firefox which I now use regularly though. I do not know how proponents of Chrome(/ium) or other browsers are able to solve this in the same manner.
```
[Match]
Name=wlan0
SSID="specificSSID"

[Network]
DHCP=ipv4
LLMNR=false
DNSOverTLS=no
DNSSEC=no
[DHCPv4]
UseDNS=false
UseDomains=false
Anonymize=true
```

That was also the reason for me to stumble across  the description of `UseDNS`.
I don't mind the description being changed instead of the application to fit the description. 
Would also be quicker.
The problem I had back then was that I was chasing a ghost, thinking my config must have some minor difference from what is expected to make it work in the described manner. With  the description being off, that undertaking is, of course, doomed.
Thanks for looking at this after such a long time.

**Comment by yuwata** [comment]
> I take the global configuration for the DNS-servers.

Hm? Where does the 'global configuration' come from? How do you configure `1.1.1.1` and `8.8.8.8`?

**Comment by teatreeoilchocolate** [comment]
> Hm? Where does the 'global configuration' come from? How do you configure 1.1.1.1 and 8.8.8.8?

I set them in the resolved.conf

```
% cat /etc/systemd/resolved.conf
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.
#
# Entries in this file show the compile time defaults.
# You can change settings by editing this file.
# Defaults can be restored by simply deleting this file.
#
# See resolved.conf(5) for details

[Resolve]
DNS=9.9.9.9
FallbackDNS=1.1.1.1 149.112.112.112
#Domains=
LLMNR=no
#MulticastDNS=yes
MulticastDNS=no
#DNSSEC=allow-downgrade
DNSSEC=allow-downgrade
#DNSOverTLS=no
DNSOverTLS=no
#Cache=yes
#DNSStubListener
```

**Comment by keszybz** [comment]
The documentation issue was resolved in b63dae3168209a06cefe2b9916667b7754b7de34. IIUC, everything else works. If the problem reappears, please open a new issue, because this one is quite messy.
